### DISCUSSION TITLE
Variable resolution

### DISCUSSION THREAD

**Discussion by adri1g** [usebruno/bruno#1469] [Q&A]
Hello community,
How can I automatically resolve variables from url, body or headers when I get them a pre or post request script using following APIs : req.getUrl(), req.getBody(), req.getHeaders().
Here an example for url : https://passport.{{host}}:{{port}}/3dpassport/login?service=https://space.{{host}}:{{port}}/3dspace
Thanks for your help,

Adrien

**Comment by helloanoop** [comment]
@adri1g Could you provide more details ?

The variables are defined in the environment and get automatically resolved before making the request.

In the URL that you shared "https://passport.{{host}}:{{port}}/3dpassport/login?service=https://space.{{host}}:{{port}}/3dspace", You will have define the `host`, `port` variables in your env vars.

**Reply by adri1g** [reply]
My need is to get a resolved url instead of the one containing variables. Is it possible ?
I'm managing url redirection in pre and post request scripts so and I dont want to resolve all variables manually.

**Reply by jthln** [reply]
@helloanoop I'm interested in such a feature, too.

Does Bruno already expose a resolver method to the user world, that allows users in their pre/post request scripts to resolve all variables contained in a given string, no matter which and how many variables of the  `collectionVariables`, `envVariables` and `processEnvVars` pools the given string contains?

**Reply by minteral** [reply]
I have the same requirement. 
In my pre-request script, I need the URL with the variables and path parameters substituted. I wrote a function to handle variable substitution, which solved part of the problem. However, I'mm still unable to substitute the path parameters.

Having the URL with substituted values would be pretty useful, if it is possible in any way.

**Comment by minteral** [comment]
This is how I have solved it for the url so far, maybe it will help:

```
// Function to substitute request variables in double curly braces
const substituteVariables = (input) => {
    return input.replace(/{{(.*?)}}/g, (match, variableName) => {
        // Get the variable value
        let value = bru.getRequestVar(variableName);
        return value !== undefined ? value : match; // If no value, keep the original placeholder
    });
};
substitutedUrl = substituteVariables(req.getUrl());
console.log('Substituted URL:', substitutedUrl);
```

**Comment by jdvivar** [comment]
This is how I solved it:
URL request looks like:
```
/service_path/v1/action/$randomUserId
```

And the pre-request script looks like:
```js
const { v4: uuidv4 } = require('uuid');
const user_id = uuidv4()
bru.setEnvVar('user_id', user_id); // we need it for future requests
req.setUrl(req.getUrl().replace("$randomUserId", user_id))
```

**Comment by brki** [comment]
There is now `bru.interpolate` :

```
const resolvedUrl = bru.interpolate(req.getUrl());
console.log("Resolved URL: " + resolvedUrl);
```
